### PR TITLE
[DEM-780] added docstring to ApiQI

### PR DESCRIPTION
**Pull request by QFer**
Added docstrings and revised docstring for all methods of Api.
Updated the unittests with some minor changes I made for getting consistency in output/interface etc.

### CODE REVIEW

**Comment by quantumkoen** [comment]
@QFer awww code coverage dropped from 100% to 96.579%, any chance we get that back up to 100%?

**Comment by QFer** [comment]
Added docstring. For some methods I added exception catching for raising an ApiError exception consistently. Unittests added for these exceptions. Furthermore type hints added for the whole file and removed types from docstrings.

File: src/quantuminspire/api.py [hunk 1]
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
...
+                | number_of_qubits (int)            | Number of qubits in the circuit for this experiment.
+                | execution_time_in_seconds (float) | The execution time of the job.
+                | raw_text (str)                    | Text string filled when an error occurred, else empty.
+                | raw_data_url (str)                | Url to get the raw data of the result. The raw data exists of a
+                |                                   | list of integer values depicting the state for each shot.
+                | histogram (OrderedDict)           | The histogram as a list of tuples with state (str) and
+                |                                   | its probability (float).
+                | histogram_url (str)               | Url to get the histogram with probabilities. This results in the
+                |                                   | OrderedDict as found in property histogram of result.
+                | measurement_mask (int)            | The measurement mask.
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]:
```suggestion
                | measurement_mask (int)            | (deprecated, unused) The measurement mask.
```

File: src/quantuminspire/api.py [hunk 2]
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
...
+                | quantum_states_url (str)          | Url to get a list of quantum states.
+                | measurement_register_url (str)    | Url to get a list of measurement register.
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]:
```suggestion
                | measurement_register_url (str)    | Url to get a list of intermediate measurement register values.
```

File: src/quantuminspire/api.py [hunk 3]
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
...
         """
-        return self._action(['results', 'list'])
-
-    def get_result(self, result_id):
-        """ Gets the histogram results of the executed QASM code.
+        try:
+            result = self._action(['results', 'read'], params={'id': result_id})
+        except Exception as ex:
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]: too broad

File: src/quantuminspire/api.py [hunk 4]
```diff
@@ -277,38 +419,70 @@ def _create_job(self, name, asset, project, number_of_shots, full_state_projecti
...
+            raise ApiError('Result with id {} does not exist!'.format(result_id))
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]: chain

File: src/quantuminspire/api.py [hunk 5]
```diff
@@ -317,40 +491,61 @@ def get_raw_data(self, result_id):
...
+                | project (str)             | Url to get the project properties for which this asset was created.
+                | project_id (int)          | The project id of the project for which this asset was created.
         """
-        return self._action(['assets', 'list'])
-
-    def get_asset(self, asset_id):
-        """ Gets the properties of an asset.
+        try:
+            asset = self._action(['assets', 'read'], params={'id': asset_id})
+        except Exception as ex:
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]: ..

File: src/quantuminspire/api.py [hunk 6]
```diff
@@ -317,40 +491,61 @@ def get_raw_data(self, result_id):
...
+            raise ApiError('Asset with id {} does not exist!'.format(asset_id))
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]: ..

File: src/quantuminspire/api.py [hunk 7]
```diff
@@ -362,46 +557,60 @@ def _create_asset(self, name, project, content):
...
 
         Returns:
-            Boolean: True if the job result could be collected else false.
+            True if the job result could be collected else False.
         """
         attempts = itertools.count() if collect_max_tries is None else range(collect_max_tries)
         for attempt in attempts:
             time.sleep(sec_retry_delay)
-            status_message = '(id {}, iteration {})'.format(quantum_inspire_job.get_job_identifier(), attempt)
+            status_message = '(id: {}, iteration: {})'.format(quantum_inspire_job.get_job_identifier(), attempt)
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]: It seems `status_message` is not used, can't we just remove it?

File: src/quantuminspire/api.py [hunk 8]
```diff
@@ -362,46 +557,60 @@ def _create_asset(self, name, project, content):
...
+            collect_tries: The number of times the status of the job is check for completion before returning.
+            default_number_of_shots: The default used number of shots for the project.
+            identifier: The identifier to generate names for the project, asset and job when necessary.
+            full_state_projection: Do not use full state projection when set to False (default is True).
 
         Returns:
-            OrderedDict: The results of the executed qasm if succesfull else an empty dictionary if
-                         the results could not be collected.
+            The results of the executed cQASM if successful else an empty dictionary if
+            the results could not be collected within the given number of tries.
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]: Refer to `get_result()` for a specification of the format of the result?

File: src/quantuminspire/api.py [hunk 9]
```diff
@@ -410,31 +619,69 @@ def execute_qasm(self, qasm, backend_type=None, number_of_shots=256, collect_tri
...
-    def execute_qasm_async(self, qasm, backend_type=None, number_of_shots=256, default_number_of_shots=256,
-                           identifier=None, full_state_projection=True, project=None, job_name=None, user_data=''):
-        """ Creates the project, asset and job with the given qasm code and returns directly without waiting
-            for the job to complete.
+    def execute_qasm_async(self, qasm: str, backend_type: Union[OrderedDict, int, str] = None,
+                           number_of_shots:int = 256, default_number_of_shots:int = 256,
+                           identifier: str = None, full_state_projection: bool = True,
+                           project: OrderedDict = None, job_name: str = None, user_data: str = '') -> QuantumInspireJob:
+        """ With this method a cQASM program (job) is scheduled to be executed asynchronously. The method returns
+            directly without waiting for the job to complete, as opposed to method 'execute_qasm' which waits for
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]:
```suggestion
            directly without waiting for the job to complete, as opposed to method `execute_qasm` which waits for
```

File: src/quantuminspire/api.py [hunk 10]
```diff
@@ -410,31 +619,69 @@ def execute_qasm(self, qasm, backend_type=None, number_of_shots=256, collect_tri
...
+            the job to finish and returns the result.
+
+            To execute a cQASM program a job is scheduled on a backend of type 'backend type' as given by the
+            parameter backend_type. Currently there are 3 backend types available:
+            1. 'QX single-node simulator'          (default for anonymous accounts)
+            2. 'QX single-node simulator SurfSara' (basic account credentials needed)
```
Comment (quantumkoen) [inline src/quantuminspire/api.py]:
```suggestion
            2. 'QX single-node simulator SurfSara' (advanced account credentials needed)
```

**Comment by quantumkoen** [comment]
Sorry, one more comment for the review: I'm getting a bunch of errors on the type hints from mypy:

```
(env) kfjmartens@ubuntu-vm:~/src/generic/SDK/src$ mypy --ignore-missing-imports -p quantuminspire
quantuminspire/api.py:124: error: "print" does not return a value
quantuminspire/api.py:261: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:308: error: "print" does not return a value
quantuminspire/api.py:351: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:370: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:427: error: "print" does not return a value
quantuminspire/api.py:472: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:489: error: Argument 1 to "get" of "QuantumInspireAPI" has incompatible type "Optional[Any]"; expected "str"
quantuminspire/api.py:490: error: Incompatible return value type (got "Union[OrderedDict[Any, Any], List[OrderedDict[Any, Any]], List[int]]", expected "List[int]")
quantuminspire/api.py:498: error: "print" does not return a value
quantuminspire/api.py:534: error: Incompatible return value type (got "OrderedDict[Any, Any]", expected "List[OrderedDict[Any, Any]]")
quantuminspire/api.py:620: error: Argument 2 to "_wait_for_completed_job" of "QuantumInspireAPI" has incompatible type "Optional[int]"; expected "int"
quantuminspire/api.py:690: error: Incompatible types in assignment (expression has type "UUID", variable has type "Optional[str]")
(env) kfjmartens@ubuntu-vm:~/src/generic/SDK/src$
```

File: src/quantuminspire/api.py [hunk 11]
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
         except Exception as ex:
             raise ApiError('Could not connect to {}'.format(base_uri)) from ex
 
-    def get(self, uri_path):
-        """ Makes a request to get request with a client instance.
+    def get(self, uri_path: str) -> Union[OrderedDict, List[OrderedDict], List[int]]:
+        """ Method for making requests to the coreapi client instance to get some piece of information. The information
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: get is used more often, also in other classes. A new ticket is oke to fix it in the whole SDK.

File: src/tests/quantuminspire/test_api.py [hunk 12]
```diff
@@ -653,7 +723,7 @@ def __fake_no_results_job_handler(self, mock_api, document, keys, params=None, v
                             ('backend', 'https,//api.quantum-inspire.com/backends/1/'),
                             ('backend_type', 'https,//api.quantum-inspire.com/backendtypes/1/'),
                             ('results', ''),
-                            ('queued_at', '2018-08-24T07,01,21.257557Z'),
+                            ('queued_at', '2018-08-24T07:01:21:257557Z'),
                             ('number_of_shots', 1)])
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/tests/quantuminspire/test_api.py]: Ok

File: src/quantuminspire/api.py [hunk 13]
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
...
+            requested depends on the uri_path parameter.
 
         Args:
-            uri_path (str): The URL where to request the data.
+            uri_path: The URL where to request the data.
+                      The uri_path has the form <api_uri><object><object-id><request>
+                      e.g. 'https://api.quantum-inspire.com/jobs/1234567/result/' to get the result of job with
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: I'll add a note for now: "This method will be made private in the near future. Usage is heavily discouraged."

[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: When making this method private, we must add all the necessary get_xxxxx methods that are needed.

File: src/quantuminspire/api.py [hunk 14]
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
...
 
         Returns:
-            OrderedDict: The properties of the backend type.
+            The requested backend type indicated by backend_id with all of its properties.
+            See 'get_default_backend_type' for the description of the backend type properties.
         """
-        return self._action(['backendtypes', 'read'], params={'id': backend_id})
+        try:
+            backend_type = self._action(['backendtypes', 'read'], params={'id': backend_id})
+        except Exception as ex:
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: I'll change it to only catch ErrorMessage. This is thrown when the result contains an error.

File: src/quantuminspire/api.py [hunk 15]
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
...
+        return self._action(['backendtypes', 'list'])
 
-    def get_backend_type_by_id(self, backend_id):
-        """ Gets the properties of a backend type.
+    def get_backend_type_by_id(self, backend_id: int) -> OrderedDict:
+        """ Gets the properties of a specific backend type, given the backend id.
 
         Args:
-            backend_id (int): The backend identification number.
+            backend_id: The backend identification number.
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: Ok

File: src/quantuminspire/api.py [hunk 16]
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
...
+                | description (str)          | Short description of the backend.
+                | topology (OrderedDict)     | Dictionary with property 'edges' (list).
+                | is_allowed (bool)          | Indicates whether the user is allowed to use this backend.
         """
-        return self._action(['backendtypes', 'list'])
+        return self._action(['backendtypes', 'default', 'list'])
 
-    def get_default_backend_type(self):
-        """ Gets the default backend type.
+    def get_backend_types(self):
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: Ok

File: src/quantuminspire/api.py [hunk 17]
```diff
@@ -48,97 +70,141 @@ def __init__(self, base_uri, authentication, project_name=None, coreapi_client_c
...
         Returns:
-            OrderedDict: The properties of the backend type.
+            The requested backend type indicated by backend_id with all of its properties.
+            See 'get_default_backend_type' for the description of the backend type properties.
         """
-        return self._action(['backendtypes', 'read'], params={'id': backend_id})
+        try:
+            backend_type = self._action(['backendtypes', 'read'], params={'id': backend_id})
+        except Exception as ex:
+            raise ApiError('Backend type with id {} does not exist!'.format(backend_id))
```
[the remark this answers is not in the document]
Comment (QFer) [reply src/quantuminspire/api.py]: Ok

File: src/quantuminspire/api.py (hunk 7 above)
Comment (QFer) [reply src/quantuminspire/api.py]: Ok, i think it was there for debugging reasons, to get feedback while waiting for the job to complete.

File: src/quantuminspire/api.py (hunk 9 above)
Comment (QFer) [reply src/quantuminspire/api.py]: Is there a convention for this?

Comment (quantumkoen) [reply src/quantuminspire/api.py]: In mark-down, back-tics are used to mark code and I always think of method names and such as code. It makes them stand out and most people will recognize it as something that refers to code.

**Comment by QFer** [comment]
I handled all comments.

**Pull request merged**